### ISSUE TITLE
Access epoch number inside custom loss function

### ISSUE THREAD

**Issue by myouness** [fchollet/keras#3155]
Hello,

I am using a graph model with one input and multiple outputs and I want to access epoch number inside a custom loss function : 

def alphabinary(alpha):
    def binary_cross(y_true, y_pred):
        return alpha \* K.mean(K.binary_crossentropy(y_pred, y_true), axis=-1)
    return binary_cross

My objective is to make the "alpha" parameter into alpha(epoch). 
I was able to do it outside the loss by recompiling but my code is kinda ugly so I am looking for a better way.
(PS: My graph model is doing target replication : Deeply-Supervised Nets )

Thank you very much for your help and have a nice day !

**Comment by evoclue** [comment]
I have the same question! It would be great if someone can share the answer to it.

**Comment by lovecambi** [comment]
I think you need Callback, and you can try this code.
`
alpha = K.variable(1.)
class NewCallback(Callback):
            def __init__(self, alpha):
                self.alpha = alpha
            
            def on_epoch_end(self, epoch, logs={}):
                K.set_value(self.alpha, K.get_value(self.alpha) * epoch**0.95)
`

**Comment by brekelma** [comment]
I have the same question!   I would like to have a loss of the form:  alpha * loss1 + loss2, with alpha changing based on epoch #.  

I have created a Callback, but its value is not dynamically updating within my custom loss.  The initial value of alpha is used throughout training.  

Is there a way to do this?!?

**Comment by JihongJu** [comment]
@lovecambi Your suggestion is way more elegant than recompiling the model, and it works. Though I think it should be `0.95 ** epoch` rather than `epoch ** 0.95`. Thank you!

**Comment by nabsabraham** [comment]
hi @JihongJu, so have you figured out how to use the callback to update your loss function based on the epoch? I'm confused as to how the alpha in the callbacks is being assigned to the alpha in your loss function. Would you mind explaining the code if possible?

**Comment by edufonseca** [comment]
hi @nabsabraham, did you finally make it work the solution by @lovecambi ? I'm also trying to implement it. The callback seems to update `self.alpha` every epoch correctly. But the update does not reach my custom loss function. Instead, `alpha` keeps the initialization value forever. 

@CVxTz how did you pass your custom `alphabinary(alpha)` loss function at compilation time? Doing as follows does not seem to work for me.

```python
alpha = K.variable(1.)
model.compile(optimizer=opt, loss=alphabinary(alpha), metrics=...)
```
thanks!

**Comment by nabsabraham** [comment]
sorry @edufonseca, i had the same question you had and could not resolve how to do this in keras. In pytorch, this is much simpler

**Comment by sdbonte** [comment]
Did anyone manage to make this work? Having a loss function which depends on the epoch would be really interesting for my research!

**Comment by ismskd** [comment]
hi@nabsabraham did u find any solution?  I also want to use the callback to update my loss function based on the epoch.

**Comment by nabsabraham** [comment]
I did not find a solution and moved to PyTorch where this is a very simple fix.

**Comment by Po-Hsuan-Huang** [comment]
This answer solved my problem.
https://stackoverflow.com/a/52713046/4755986

**Comment by edufonseca** [comment]
It works for me using a callback and a function closure. Check the repo below for working code.

callback:
https://github.com/edufonseca/waspaa19/blob/8c5dd31922911be861f7c5f4dcc7eec058c8ad67/main.py#L464

compile model within a class:
https://github.com/edufonseca/waspaa19/blob/8c5dd31922911be861f7c5f4dcc7eec058c8ad67/loss_time.py#L35

function closure:
https://github.com/edufonseca/waspaa19/blob/8c5dd31922911be861f7c5f4dcc7eec058c8ad67/loss_time.py#L123

hth

**Comment by ssavale-entrata** [comment]
[lovecambi](https://github.com/lovecambi)'s solution is good, but it doesn't work with Keras hyper parameter tuning.

**Comment by Faptimus420** [comment]
After spending days on this myself, I found a mind-bendingly simple method for doing this that is fully compatible with Keras Tuner as well, and doesn't require using Callbacks at all. I've used the backend-agnostic Keras Core Variable class in my example, but it should work just as well with the old Keras/TF Variables.

- In your model's __init__() method, define the initial state of the non-trainable variable:
`self.current_epoch = Variable(initializer=0, dtype='int32', trainable=False)`
- In the model's call() method, increment the current_epoch variable by 1 - the call() method is called on every epoch, so it takes care of the incrementation on every epoch. If you assign_add as the first thing within call(), the variable will be updated at the beginning of the epoch, if you do it as the last thing, it will be updated at the end of the epoch.
`self.current_epoch.assign_add(1)`
- I needed the epoch number to implement custom weight regularization, but if you need it for the loss function fore example, within the train_step() method, just call your loss like this:
`loss = compute_loss(y, y_pred, self.current_epoch.value)`
